### PR TITLE
Display local/completed transactions

### PR DESCRIPTION
**Pull request by jacogr**
- TxList component for transaction display (split from Account/Transactions)
- Display local transactions under signer page

Normal transaction list -

parity 2016-11-27 12-13-57

Signer list -

parity 2016-11-27 12-14-26

### CODE REVIEW

**Comment by ngotchac** [comment]
47 years ago

screenshot_2016-11-28_13-30-52

File: js/src/ui/TxList/store.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+        this.loadBlocks(transactions.map((tx) => tx.blockNumber.toNumber()));
+      })
+      .catch((error) => {
+        console.warn('loadTransactions', error);
+      });
+  }
+
+  loadBlocks (_blockNumbers) {
+    const blockNumbers = Object.keys(
```
Comment (ngotchac) [inline js/src/ui/TxList/store.js]: Lodash `uniq` method on requested blocks?

File: js/src/ui/TxList/store.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        .sort((ahash, bhash) => {
+          const bnA = this.transactions[ahash].blockNumber;
+          const bnB = this.transactions[bhash].blockNumber;
+
+          return bnB.comparedTo(bnA);
+        });
+    });
+  }
+
+  loadTransactions (_txhashes) {
```
Comment (ngotchac) [inline js/src/ui/TxList/store.js]: This should be an `@action` right ?

File: js/src/views/Signer/store.js [hunk 3]
```diff
@@ -63,4 +80,20 @@ export default class Store {
         console.warn('Store:fetchBalances', error);
       });
   }
+
+  fetchLocalTransactions = () => {
+    const nextTimeout = () => {
+      if (this.doPolling) {
+        setTimeout(this.fetchLocalTransactions, 1500);
```
Comment (ngotchac) [inline js/src/views/Signer/store.js]: Might be safer to add `clearTimeout` with the timeout id no?

**Comment by ngotchac** [comment]
Looks good except the wrong date (when not mined yet) and a few small comments :)

File: js/src/ui/TxList/store.js (hunk 2 above)
Comment (jacogr) [inline js/src/ui/TxList/store.js]: Could be. Technically the only updates happen in the `addTransactions` function. So `@action` won't hurt, however the function itself doesn't update the object state. (Needed to split the change out into a different `@action` - async, so doesn not play well with `setStrict`)

File: js/src/ui/TxList/store.js (hunk 1 above)
Comment (jacogr) [inline js/src/ui/TxList/store.js]: Cool.

File: js/src/ui/TxList/store.js (hunk 2 above)
Comment (ngotchac) [inline js/src/ui/TxList/store.js]: Mh that's right, didn't think of that. Not sure what it changes.

**Comment by jacogr** [comment]
Next PR will cover MethodDecoding (pre-confirmation), gas edits & re-submissions.

File: js/src/ui/TxList/txList.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    api: PropTypes.object.isRequired
+  }
+
+  static propTypes = {
+    address: PropTypes.string.isRequired,
+    hashes: PropTypes.oneOfType([
+      PropTypes.array,
+      PropTypes.object
+    ]).isRequired,
+    isTest: PropTypes.bool.isRequired
```
Comment (ngotchac) [inline js/src/ui/TxList/txList.js]: `isTest` is required but not provided. I think it should be removed from here.

File: js/src/views/Signer/containers/RequestsPage/RequestsPage.js [hunk 5]
```diff
@@ -44,39 +46,59 @@ class RequestsPage extends Component {
     isTest: PropTypes.bool.isRequired
```
Comment (ngotchac) [inline js/src/views/Signer/containers/RequestsPage/RequestsPage.js]: Same as above

File: js/src/ui/TxList/txList.js (hunk 4 above)
Comment (jacogr) [inline js/src/ui/TxList/txList.js]: Provided via mapStateToProps - however we do have an issue with the new 'undefined' state now. (In multiple places)

**Comment by ngotchac** [comment]
Looks good overall. The Local and Finished transactions seems redundant (we already discussed about it, just wanted to write it down)

File: js/src/ui/TxList/txList.js (hunk 4 above)
Comment (ngotchac) [inline js/src/ui/TxList/txList.js]: Ah yep that's right... So maybe it's not _required_ ? Or it could be defaulted to false in `mapStateToProps`

**Comment by jacogr** [comment]
Agreed on the Finished. Only benefit now is showing the Rejected transactions, as soon as we have resubmissions in place, it should be dropped and combined with the Local - that was it makes sense, can reject and re-submit again.

File: js/src/ui/TxList/txList.js (hunk 4 above)
Comment (jacogr) [inline js/src/ui/TxList/txList.js]: Well, it is required to be provided - if we don't have it, we will get the links wrong. The issue however is that it turned into a tri-state at this point as opposed to the original bool and the isRequired doesn't see it as such, not can it. (We need a different solution for that problem, it is not local to this component - sadly.)

Comment (ngotchac) [inline js/src/ui/TxList/txList.js]: Agreed, not in the scope of this issue. Should be null/true/false and the prop type should be a nullable boolean I guess.

**Pull request merged**